### ISSUE TITLE
jQuery ajax request handles 302 or 307 Temporary Redirect as error

### ISSUE THREAD

**Issue by TuningGuide** [jquery/jquery#2402]
I did something like this:

$.getJSON('/tools/checkifemailexists.php?email='+email, function() { console.log("success")}).fail(function() { console.log("error")}).

When my php server then answers with a Temporary Redirect. My browser does the redirect directly and answers with the expected response. But jquery handles the response as error. Is this a bug or intended behaviour?

**Comment by dmethvin** [comment]
Did the response return valid JSON?

**Comment by TuningGuide** [comment]
Yep, the response was:
{
    "id": "119XXXX",
    "type": "user"
}

**Comment by dmethvin** [comment]
It would really help to have a test case to debug this, or you can debug it and let us know where it is failing.

**Comment by TuningGuide** [comment]
Do you know any debuggin platform where I can force a 307 / 302 redirect?

**Comment by dmethvin** [comment]
Unfortunately, no. You'd need to create a test and run it locally on your own server to be sure there aren't any conditions on the server that are causing this. For the case of a 302 the browser should be following the redirect and jQuery won't even see it, just the final URL. The browser should be doing the same with a 307, meaning jQuery doesn't even know it's a redirect.

**Comment by mzgol** [comment]
> Do you know any debuggin platform where I can force a 307 / 302 redirect?

I don't know about 307 but you can force 302-redirects via http://httpbin.org/. A very handy service for testing cross-origin stuff. :-)

**Comment by TuningGuide** [comment]
Ok, it seems to have something to do with the header field: "X-Requested-With: XMLHttpRequest". My playground is: http://jsbin.com/seqinupumi/edit?js,output. On my server it adds this field and then does not do a redirect but gets the content directly within the one request. If I add crossDomain: true to my ajax options. It does the redirect and jquery will output "success"

**Comment by TuningGuide** [comment]
@dmethvin why did you close the ticket! The underlying Problem is not solved for me. I just found a quirky workaround.

**Comment by dmethvin** [comment]
Oh, I thought you were saying you understood what happened. It sounds like your local server redirected to a remote resource, but as a result the request was cross-domain and hadn't been preflighted. If you say it's cross-comain, which isn't something jQuery can know, the request succeeds?

**Comment by TuningGuide** [comment]
Thats where the magic does not happen (my status code is 302):

			if ( status >= 200 && status < 300 || status === 304 ) {

				// Set the If-Modified-Since and/or If-None-Match header, if in ifModified mode.
				if ( s.ifModified ) {

					modified = jqXHR.getResponseHeader("Last-Modified");
					if ( modified ) {
						jQuery.lastModified[ ifModifiedKey ] = modified;
					}
					modified = jqXHR.getResponseHeader("Etag");
					if ( modified ) {
						jQuery.etag[ ifModifiedKey ] = modified;
					}
				}

				// If not modified
				if ( status === 304 ) {

					statusText = "notmodified";
					isSuccess = true;

				// If we have data
				} else {

					isSuccess = ajaxConvert( s, response );
					statusText = isSuccess.state;
					success = isSuccess.data;
					error = isSuccess.error;
					isSuccess = !error;
				}
			} else {
				// We extract error from statusText
				// then normalize statusText and status for non-aborts
				error = statusText;
				if ( !statusText || status ) {
					statusText = "error";
					if ( status < 0 ) {
						status = 0;
					}
				}
			}

**Comment by TuningGuide** [comment]
@dmethvin Could you not at least reopen this ticket as the error is still present. Regards

**Comment by mzgol** [comment]
I'll reopen for now but please create a full minimal test case reproducing the problem on https://jsbin.com. You can use http://httpbin.org/ for testing response handling.

**Comment by TuningGuide** [comment]
@mzgol  Hey thank you. I took your advice and created a test case. But I had to do it on my own webspace as it would have been a cross-domain request otherwise. But there I figured out that the framework we use (CakePHP) was doing the redirect not with the Location header field but somehow internally. Therefore it did not do the second request separately and jquery did not know how to handle the 307 without a following 200. Anyway thanks for reopening but now it can be closed xD
